include/grpc/impl/codegen/sync_posix.h

```diff
@@ -25,8 +25,26 @@
 
 #include <pthread.h>
 
+#ifdef GRPC_ASAN_ENABLED
+/* The member |leak_checker| is used to check whether there is a memory leak
```

Comment:
Instead of changing the structure and all the methods due for ASAN and non-ASAN builds, I would try using  `__lsan_register_root_region()` and `__lsan_unregister_root_region()` first:
https://github.com/llvm-mirror/compiler-rt/blob/master/include/sanitizer/lsan_interface.h#L29-L41

Basically, we would need to define macros for this:
```
#if ASAN_ENABLED
#define LSAN_REGISTER_ROOT_REGION(x, s) __lsan_register_root_region(x, s)
#define LSAN_UNREGISTER_ROOT_REGION(x, s) __lsan_unregister_root_region(x, s)
#else 
#define LSAN_REGISTER_ROOT_REGION(x, s)
#define LSAN_UNREGISTER_ROOT_REGION(x, s)
#endif
```

I believe this should work. If that didn't pan out, using the current approach would be fine.

Reply:
Interesting, I wasn't aware of this. So this is basically like the current approach, but it won't require changing the member variables of the gpr_mu struct itself? And we'd do LSAN_REGISTER_ROOT_REGION in gpr_mu_init and LSAN_UNREGISTER_ROOT_REGION in gpr_mu_destroy?

Reply:
Yes exactly, malloc/free has these annotations. For memory that's mmaped or inside another heap-allocated structure, we just need to tell ASAN, that the cv and mu structures are leak roots, and have to be unregistered.

Reply:
Thanks for your review. It's worth trying.

Reply:
@soheilhy I found there's a [leak_checker](https://github.com/abseil/abseil-cpp/blob/cc4bed2d74f7c8717e31f9579214ab52a9c9c610/absl/debugging/leak_check.cc#L26) has wrapped the LSAN interfaces under the MACRO `LEAK_SANITIZER`. I'm a little confused that these test codes are written under the MACRO `GRPC_ASAN_ENABLED`, should we move the codes to LSAN region?

Reply:
AFAIK, that's historical. Leak detection a bit older than address sanitizer, and historically everything is prefixed as LSAN. The leak detection part of ASAN relies on LSAN: 
https://clang.llvm.org/docs/AddressSanitizer.html#suppressing-memory-leaks

So when `GRPC_ASAN_ENABLED` is true, we have LSAN implicitly. 

Some frameworks use LSAN without ASAN, but for gRPC I think we run TSAN, ASAN, and MSAN. So, for gRPC I think guarding under ASAN or LSAN would be both fine. Anything that works would be fine :-)

Reply:
@soheilhy I've changed to using `__lsan_register_root_region()` and `__lsan_unregister_root_region()`. And reverted two leaks' fixing to check whether the test suits can catch the leaks.

Reply:
Reverted.